handle state tensors in training ir path

avikchaudhuri opened the pull request:
Summary: We had attribute assignment detection and handling of registered buffer assignments when using `aot_autograd`, but not when using just `make_fx`. Fixed.

Test Plan: expanded coverage of `test_state_tensors` to use `export` instead of `torch.export.export`

Differential Revision: D63802576

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1423,12 +1426,57 @@ def _make_fx_helper(mod, args, kwargs, **flags):
+                        if node.name == name:
+                            add_nodes.append(node)
+                            node.users[output_node] = None
+                output_node.args = ((*add_nodes, *output_node.args[0]),)
```
In training IR, we are following the convention that we don't return mutated buffers. Could we just write it as inplace update?

avikchaudhuri replied:
If the source code has `buf = new_value` can we treat it the same as `buf.copy_(new_value)`?

tugsbayasgalan replied:
I think this is ok as long as metadata etc remains the same (aka shape). I mean when we unlift, we just write them as copy_ anyways... So seems ok? What do you think?

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1423,12 +1426,57 @@ def _make_fx_helper(mod, args, kwargs, **flags):
+                output_node = None
+                output_node = list(gm.graph.nodes)[-1]
+                for buf, name in assigned_buffers.items():  # type: ignore[possibly-undefined]
+                    buffers_to_mutate[name] = buf
```
Same here: Can we make it a helper function in aot_autograd and use it here?

avikchaudhuri replied:
This part is going to be substantially different, adding copy nodes instead of mutating buffer. I'll dedup the other part.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1423,12 +1426,57 @@ def _make_fx_helper(mod, args, kwargs, **flags):
+
+                assigned_buffers = {}
+
+                def _map_assigned_buffer_to_proxy(_mod, name, buffer):
```
Could we make this a util function in aot_autograd.py and use that here? I think this logic also exist in aot_autograd.py?

avikchaudhuri replied:
Mostly true, although there we are seeing buffers as Functional(Fake(...)) and here just Fake(...).

tugsbayasgalan replied:
Could we filter out the other parts? What i am worried is this part is bit complicated to understand so it would be better if we have only place to change stuff. Maybe the util function can take what instance we want to assert on lol.

avikchaudhuri committed:
handle state tensors in training ir path (#137240)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/137240

We had attribute assignment detection and handling of registered buffer assignments when using `aot_autograd`, but not when using just `make_fx`. Fixed.

Test Plan: expanded coverage of `test_state_tensors` to use `export` instead of `torch.export.export`

Differential Revision: D63802576
